olive/passes/olive_pass.py

```diff
@@ -139,14 +139,15 @@ def default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConf
             config.update(get_user_script_config())
         # add all other parameters
         config.update(cls._default_config(accelerator_spec))
-        # validate that all parameters ending with data_config are of type DataConfig, Union[DataConfig, dict], ...
+        # validate that all parameters whose ParamCategory is DATA_CONFIG
+        # are of type DataConfig, Union[DataConfig, dict], ...
         # this requirement is on the pass developer but we can only check it here
         for param, param_config in config.items():
-            if param.endswith("data_config"):
+            if param_config.category == ParamCategory.DATA_CONFIG:
                 param_type = param_config.type_
                 assert param_type == DataConfig or DataConfig in get_args(
                     param_type
-                ), f"{param} ending with data_config must be of type DataConfig."
+                ), f"{param} must be of type DataConfig."
```

Comment:
If we have data configs at system level, we can resolve data config when before the pass run. Do we have any other use case for the nested data config in pass config?

Reply:
The data configs are pass config parameters. So, we need to know what parameters data configs are so that we can resolve string names in the run config validation as well as perform some of the model name and task auto insertion.

Reply:
I added a new `ParamCategory.DATA_CONFIG` so we don't need to identify data config param by its name. If it is a dict, this step might be a little hacky.

Reply:
the paramcategory is only helpful inside the aml or docker systems since the pass is already instantiated with `accerator_spec`. For the workflow config json validation, we don't have access to the instantiated pass config so we don't know which params are data configs without some way. That's why after discussion with @guotuofeng, we decided to reserve `data_config` suffix for such params. https://github.com/microsoft/Olive/blob/main/olive/workflows/run/config.py#L204

Dicts are fine since the dict gets converted into a DataConfig object after validation by pydantic anyways. We can also remove `dict` from the `type_`. It was added to make it clear we can provide a dict as value instead but internally, since DataConfig is first in the Union, all dicts converted to DataConfig.

Reply:
I see. I will revert this change for now, and will investigate more on this.